File: ciam/src/main/java/com/crm/ciam/security/JwtService.java
package com.crm.ciam.security;

import java.security.KeyPair;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.util.Date;
import java.util.List;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import com.crm.ciam.entity.Benutzer;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;

@Service
public class JwtService {

    private final RSAPrivateKey privateKey;
    private final RSAPublicKey publicKey;
    private final long accessTokenExpiration;

    public JwtService(
            KeyPair rsaKeyPair,
            @Value("${jwt.access-token-expiration}") long accessTokenExpiration) {
        this.privateKey = (RSAPrivateKey) rsaKeyPair.getPrivate();
        this.publicKey = (RSAPublicKey) rsaKeyPair.getPublic();
        this.accessTokenExpiration = accessTokenExpiration;
    }

    public String generateAccessToken(Benutzer benutzer) {
        List<String> rollen = benutzer.getRollen().stream()
                .map(Enum::name)
                .toList();

        List<String> permissions = RolePermissionMapping.getPermissions(benutzer.getRollen()).stream()
                .map(Permission::name)
                .toList();

        return Jwts.builder()
                .subject(benutzer.getBenutzername())
                .claim("rollen", rollen)
                .claim("permissions", permissions)
                .claim("vorname", benutzer.getVorname())
                .claim("nachname", benutzer.getNachname())
                .claim("benutzerId", benutzer.getId())
                .issuedAt(new Date())
                .expiration(new Date(System.currentTimeMillis() + accessTokenExpiration))
                .signWith(privateKey, Jwts.SIG.RS256)
                .compact();
    }

    public String extractUsername(String token) {
        return extractAllClaims(token).getSubject();
    }

    public boolean validateToken(String token) {
        try {
            extractAllClaims(token);
            return true;
        } catch (JwtException | IllegalArgumentException e) {
            return false;
        }
    }

    public Claims extractAllClaims(String token) {
        return Jwts.parser()
                .verifyWith(publicKey)
                .build()
                .parseSignedClaims(token)
                .getPayload();
    }
}

File: ciam/src/main/kotlin/com/crm/ciam/security/JwtService.kt
package com.crm.ciam.security

import com.crm.ciam.entity.Benutzer
import io.jsonwebtoken.Claims
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.security.KeyPair
import java.security.interfaces.RSAPrivateKey
import java.security.interfaces.RSAPublicKey
import java.util.Date

@Service
class JwtService(
    rsaKeyPair: KeyPair,
    @Value("\${jwt.access-token-expiration}") private val accessTokenExpiration: Long
) {

    private val privateKey: RSAPrivateKey = rsaKeyPair.private as RSAPrivateKey
    private val publicKey: RSAPublicKey = rsaKeyPair.public as RSAPublicKey

    fun generateAccessToken(benutzer: Benutzer): String {
        val rollen = benutzer.rollen.map { it.name }
        val permissions = RolePermissionMapping.getPermissions(benutzer.rollen).map { it.name }

        return Jwts.builder()
            .subject(benutzer.benutzername)
            .claim("rollen", rollen)
            .claim("permissions", permissions)
            .claim("vorname", benutzer.vorname)
            .claim("nachname", benutzer.nachname)
            .claim("benutzerId", benutzer.id)
            .issuedAt(Date())
            .expiration(Date(System.currentTimeMillis() + accessTokenExpiration))
            .signWith(privateKey, Jwts.SIG.RS256)
            .compact()
    }

    fun extractUsername(token: String): String =
        extractAllClaims(token).subject

    fun validateToken(token: String): Boolean =
        try {
            extractAllClaims(token)
            true
        } catch (e: JwtException) {
            false
        } catch (e: IllegalArgumentException) {
            false
        }

    fun extractAllClaims(token: String): Claims =
        Jwts.parser()
            .verifyWith(publicKey)
            .build()
            .parseSignedClaims(token)
            .payload
}
